core/reorder/mc64.hpp

```diff
@@ -0,0 +1,112 @@
+        array<remove_complex<ValueType>>& weights_array,          \
+        array<remove_complex<ValueType>>& dual_u_array,           \
+        array<remove_complex<ValueType>>& row_maxima_array,       \
+        gko::experimental::reorder::mc64_strategy strategy)
+
+#define GKO_DECLARE_MC64_INITIAL_MATCHING(ValueType, IndexType)              \
+    void initial_matching(                                                   \
+        size_type num_rows, const IndexType* row_ptrs,                       \
+        const IndexType* col_idxs, const array<ValueType>& weights_array,    \
+        const array<ValueType>& dual_u_array, array<IndexType>& permutation, \
+        array<IndexType>& inv_permutation,                                   \
+        array<IndexType>& matched_idxs_array,                                \
+        array<IndexType>& unmatched_rows_array, ValueType tolerance)
+
+#define GKO_DECLARE_MC64_SHORTEST_AUGMENTING_PATH(ValueType, IndexType)   \
+    void shortest_augmenting_path(                                        \
+        size_type num_rows, const IndexType* row_ptrs,                    \
+        const IndexType* col_idxs, array<ValueType>& weights_array,       \
+        array<ValueType>& dual_u_array, array<ValueType>& distance_array, \
+        array<IndexType>& permutation, array<IndexType>& inv_permutation, \
+        IndexType root, array<IndexType>& parents_array,                  \
+        array<IndexType>& generation_array,                               \
+        array<IndexType>& marked_cols_array,                              \
+        array<IndexType>& matched_idxs_array,                             \
+        addressable_priority_queue<ValueType, IndexType>& queue,          \
+        std::vector<IndexType>& q_j, ValueType tolerance)
+
+#define GKO_DECLARE_MC64_COMPUTE_SCALING(ValueType, IndexType)              \
+    void compute_scaling(                                                   \
+        const matrix::Csr<ValueType, IndexType>* mtx,                       \
+        const array<remove_complex<ValueType>>& weights_array,              \
+        const array<remove_complex<ValueType>>& dual_u_array,               \
+        const array<remove_complex<ValueType>>& row_maxima_array,           \
+        const array<IndexType>& permutation,                                \
+        const array<IndexType>& matched_idxs_array, mc64_strategy strategy, \
+        ValueType* row_scaling, ValueType* col_scaling)
+
+
+template <typename ValueType, typename IndexType>
+GKO_DECLARE_MC64_INITIALIZE_WEIGHTS(ValueType, IndexType);
+
+template <typename ValueType, typename IndexType>
+GKO_DECLARE_MC64_INITIAL_MATCHING(ValueType, IndexType);
+
+template <typename ValueType, typename IndexType>
+GKO_DECLARE_MC64_SHORTEST_AUGMENTING_PATH(ValueType, IndexType);
+
+template <typename ValueType, typename IndexType>
+GKO_DECLARE_MC64_COMPUTE_SCALING(ValueType, IndexType);
```

Comment:
do you need to declare the template here?

Reply:
Yes, because I want to make sure the functions are available in the header for testing. The issue why I needed theses macros at all is that if template functions only get instantiated inside a function with internal linkage (anonymous namespace), their symbols may not get exported if the function is inlined. So I need to explicitly instantiate them.